util/pkg/vfs/s3context.go

```diff
@@ -42,7 +42,7 @@ var (
 	// https://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region
 	// TODO: perhaps make region regex more specific, i.e. (us|eu|ap|cn|ca|sa), to prevent matching bucket names that match region format?
 	//       but that will mean updating this list when AWS introduces new regions
-	s3UrlRegexp = regexp.MustCompile(`s3([-.](?P<region>\w{2}-\w+-\d{1})|[-.](?P<bucket>[\w.\-\_]+)|)?.amazonaws.com(.cn)?(?P<path>.*)?`)
+	s3UrlRegexp = regexp.MustCompile(`(s3([-.](?P<region>\w{2}-\w+-\d{1})|[-.](?P<bucket>[\w.\-\_]+)|)?|(?P<bucket>[\w\-\_]+).s3.(?P<region>\w{2}-\w+-\d{1})).amazonaws.com(.cn)?(?P<path>.*)?`)
```

Comment:
Why excluding '.' in bucket names in virtual hosted style?

Reply:
Having a dot in the bucket name causes https connection to fail:

https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html#bucketnamingrules

> When you use virtual hosted–style buckets with Secure Sockets Layer (SSL), the SSL wildcard certificate only matches buckets that don't contain periods. To work around this, use HTTP or write your own certificate verification logic. We recommend that you do not use periods (".") in bucket names when using virtual hosted–style buckets.

While mirroring assets should work (as kops uses an aws sdk s3 client), the user data script is likely to fail as kops tries to fetch assets from `https://bucket`

Reply:
That would be a reason to avoid writing out virtual hosted style URIs for such buckets, but doesn't seem a reason to avoid recognizing such URIs.

Reply:
Ah - there's actually a bug underlying this difference of opinion :-)  VFSPath is used in two different places for two quite different things.

Can I suggest that we rename VFSPath to S3URLForHTTPURL or similar? (If you can find something that doesn't look like shouting in the go naming rules, that's even better!)

I started going down this little rabbit hole and it looks like we actually need a bigger refactor... I think one of the usages (storage.go) is really `ParseS3Path` and the other one (copyfile.go) is really MapToUploadLocation (and should probably just be inlined in that function.

I'm happy to do that refactor if you want, but I don't want to "steal" the PR from you two that have actually discovered the problem :-)

Reply:
@justinsb thanks for the feedback, I did a small refactoring, but not entirely sure whether it goes in the imaging direction.

Reply:
Scratch that, I reverted my attempt as I was starting to duplicate a lot of logic and in the end those functions are fairly similar.

Instead I added the `.` as asked for by @johngmyers and as a customer your could also specify `http://my-bucket`, which would still work, but is less secure obviously.